aptos-move/framework/move-stdlib/src/natives/reflect.rs

```diff
@@ -0,0 +1,84 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implementation of native functions for reflection.
+
+use crate::natives::result;
+use aptos_gas_schedule::gas_params::natives::aptos_framework::REFLECT_RESOLVE_BASE;
+use aptos_native_interface::{
+    safely_pop_arg, safely_pop_type_arg, RawSafeNative, SafeNativeBuilder, SafeNativeContext,
+    SafeNativeError, SafeNativeResult,
+};
+use move_core_types::{
+    account_address::AccountAddress, identifier::Identifier, language_storage::ModuleId,
+};
+use move_vm_runtime::native_functions::NativeFunction;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    values::{Struct, StructRef, Value, VectorRef},
+};
+use smallvec::{smallvec, SmallVec};
+use std::{collections::VecDeque, iter};
+
+const INVALID_IDENTIFIER: u16 = 0;
+
+fn native_resolve(
+    context: &mut SafeNativeContext,
+    mut ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    // Charge base cost before anything else.
+    context.charge(REFLECT_RESOLVE_BASE)?;
+
+    // Process arguments
+    debug_assert!(ty_args.len() == 1);
+    let fun_ty = safely_pop_type_arg!(ty_args);
+    debug_assert!(args.len() == 3);
+    let Some(fun_name) = identifier_from_string(safely_pop_arg!(args))? else {
+        return Ok(smallvec![result::err_result(pack_err(INVALID_IDENTIFIER))]);
+    };
+    let Some(mod_name) = identifier_from_string(safely_pop_arg!(args))? else {
+        return Ok(smallvec![result::err_result(pack_err(INVALID_IDENTIFIER))]);
+    };
+    let addr = safely_pop_arg!(args, AccountAddress);
+    let mod_id = ModuleId::new(addr, mod_name);
+
+    // Resolve function and return closure. Notice the loader context function
+    // takes car of gas metering and type checking.
+    match context
+        .loader_context()
+        .resolve_function(&mod_id, &fun_name, &fun_ty)?
+    {
+        Ok(fun) => {
+            // Return as a closure with no captured arguments
+            Ok(smallvec![result::ok_result(Value::closure(
+                fun,
+                iter::empty()
+            ))])
+        },
+        Err(e) => Ok(smallvec![result::err_result(pack_err(e as u16))]),
```

Comment:
What if error is due to gas charging? We cannot remap it

Reply:
I also needed to look a while.

In this case `resolve_function(..)?` would fail. The errors returned here are 1:1 mapping to the enum in Move.